packages/flutter/lib/src/foundation/node.dart

```diff
@@ -115,25 +116,25 @@ class AbstractNode {
   }
 
   /// The parent of this node in the tree.
-  AbstractNode? get parent => _parent;
-  AbstractNode? _parent;
+  T? get parent => _parent;
+  T? _parent;
```

Comment:
I don't think that's right. There's no way for the class to know what it's _parent_ type will be (and it certainly doesn't need to be the same as the child type). RenderObject in particular is happy to have a parent that's not a RenderObject.

Reply:
If the performance improvement here is significant, we could make an explicit decision to change this design, and require that RenderObject trees not have a `parent` pointing to a non-RenderObject parent. If we do that we should update documentation and logic. Probably need to carefully read every mention of `parent` at least in `object.dart`, maybe other places like `binding.dart`, `view.dart`, and maybe `widgets/framework.dart`.

Reply:
cc @goderbauer for thoughts on making such a radical change to AbstractNode's design.

Reply:
Yeah, that is the unfortunate part about this PR - to remove the cast from parent we'd need to enforce that each AbstractTree subtype knows its parent type.

Reply:
while I'm at it ... I wonder If I could also enforce that parent is non-null except for root.

Reply:
> thoughts on making such a radical change to AbstractNode's design.

We'd have to to an audit, but I don't have any immediately obvious concerns.

It does seem to clean up the code a little bit. I do wonder, though, how significant this change is for performance.

Reply:
We walk parent render objects in a few places, finding out the improvement would rely on finding a scenario where those functions are used heavily.

Reply:
i.e. getTransformTo, scroll offset reveal, et cetera